### ISSUE TITLE
local_copy_list: nodes vs files

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida_core#2709]
Currently, the `local_copy_list` wants as the first argument the uuid of a node.
This means it is possible that a node has to be created just in order to be able to be passed to the local copy list.
Is this intended? What happens to these nodes, are they somehow linked to the calculation?

For example, one of my plugins used to create an input file from a set of input parameters.
This now has to be wrapped in a node.
And since it is not possible to create a Singlefiledata from a string, it actually needs to be written to disk twice (once as a file, then when the Singlefiledata is copying it).

So, in a sense, these are two questions:
 1. Should the `local_copy_list` be limited to files stored in nodes?
 2. Should one be able to construct a `SinglefileData` from a string?

P.S. An intermediate solution for 2. could be to allow a `StringIO` wrapper passed to the `file` argument of the constructor. 
Currently, SinglefileData will do `file.name`, which `StringIO` does not provide.
Once could try `file.name` and if it is not implemented, give a random (or default) name.

**Comment by ltalirz** [comment]
mentioning @sphuber for comment

**Comment by sphuber** [comment]
Your last point is a good one. I recently changed the interface of `SinglefileData` to accept both a filepath or a filelike object exactly for the reason you describe. A stream should be perfectly acceptable. If it chokes on a `.name` call that is a bug and we should fix it, probably as you suggest with some default name.

The rest of your post seems to be a bit of an XY-problem, which stems from the fact that the intended use of the `folder` and `local_copy_list` are not very clear.

In a `CalcJob` one has two ways of using the input nodes that it received to create the input data for the actual calculation. If the inputs contain a `SinglefileData` or `FolderData` who have files that need to be copied to the working directory wholesale, this is where you would use the `local_copy_list`. The nodes passed in here, will not be linked in any way, because they are already input nodes, so they will have an input link, it is just an easy way to specify which files need to be copied where on the remote. Note that this does not just work for `SinglefileData` but in principle for any data node, which is why the second argument is the relative filepath within the repository, because it should not necessarily be a single file as in the case of `SinglefileData`.

Now in all other cases, where the input is actually "composed" based on the contents of one or multiple input nodes, you would simply use the input "folder" that is passed in to the `prepare_for_submission` method by the engine. Since this is just a folder, this gives optimal freedom to build and write the input files necessary however the developer wants. Since these files are composed, the contents of this folder is stored in the `raw_input` folder of the repository of the calculation node as additional "provenance".

TL:DR; In your example it seems that you are writing an input files that is composed from the contents of multiple inputs nodes. In this case, you should simply directly write the file to the `folder` that is passed into the function.

I think the solution for this problem is simply that it needs better documentation.

**Comment by ltalirz** [comment]
@sphuber thanks for the helpful explanation.

Indeed 1. is solved by using the folder directly using something like:
```python
        with io.StringIO(configure.configure_string) as handle:
            folder.create_file_from_filelike(handle, filename=DEFAULT_CONFIGURE_FILE_NAME, mode='w')
```
I've just added this to the plugin migration guide.

2. is addressed by PR #2722
